man/systemd.unit.xml

```diff
@@ -1792,6 +1792,15 @@
           into.</para>
           </listitem>
         </varlistentry>
+
+        <varlistentry>
+            <term><varname>MaxInstances=</varname></term>
+
+            <listitem><para>The maximum number of instances of a template unit that may run simultaneously. If reached,
+            further attempts to start another instance of the same template will fail. Defaults to
```

Comment:
Just for clarification: Looking at manpages and tests for this change (I honestly did not even try to understand the implementation), units started after exceeding this limit will just move to a failed state? Or will the units be blocked and queued in a starting/ waiting state, until one of the running instances terminates?

The issue author of https://github.com/systemd/systemd/issues/17097 (which this claims to resolve) writes about currently using `flock` for this purpose. As `flock` can operate in a blocking (processes wait until they can acquire a lock) or in a non-blocking (if lock is not available, `flock` exits with a failure) way, it is not entirely clear which behaviour is desired there.
But for resolving a similar feature request https://github.com/systemd/systemd/issues/28075, it is required to in the end run all the started instances at some point, but just enqueue and block those exceeding a concurrency limit.

I'd welcome a clarification on this, maybe we could even support both modes of operation here?

Reply:
So, this current PR makes the activation fail. Which is the behaviour I think makes sense by default, simply as a mechanism to put limits on things.

The other use-case makes sense too though, but I think should be opt-in since it might result in unexpected deadlocks, since progress cannot necessarily be made if this mode is enabled for arbitrary services.